dopamine/discrete_domains/run_experiment.py

```diff
@@ -600,17 +609,57 @@ def _run_iterations(self):
     time an iteration completes a new one starts until the right number of
     iterations is run.
     """
-    # TODO(aarg): Change the thread management to an implementation with queues.
-    threads = []
+    # TODO(aarg): Change the thread management to an implementation with queues
+    # and with a fix number of thread workers. With the revamp, training might
+    # go inline to this method.
+    training_worker = threading.Thread(target=self._run_training_steps)
+    training_worker.start()
+    experience_threads = []
     for iteration in range(self._start_iteration, self._num_iterations):
-      self._running_iterations.acquire()
+      self._experience_queue.put(1)
       thread = threading.Thread(
           target=self._run_one_iteration, args=(iteration,))
       thread.start()
-      threads.append(thread)
-    # Wait for all running iterations to complete.
-    for thread in threads:
+      experience_threads.append(thread)
+
+    # Wait for all tasks to complete.
+    self._experience_queue.join()
+    self._training_queue.join()
+    # Indicate training step thread to stop.
+    self._training_queue.put(None)
+    # Wait for all running threads to complete.
+    for thread in experience_threads:
       thread.join()
+    training_worker.join()
+
+  def _begin_episode(self, observation):
+    # Increments training steps and blocks if training is too slow.
+    self._enqueue_training_step()
+    return self._agent.begin_episode(observation, training=False)
+
+  def _step(self, reward, observation):
+    # Increments training steps and blocks if training is too slow.
+    self._enqueue_training_step()
+    return self._agent.step(reward, observation, training=False)
+
+  def _enqueue_training_step(self):
+    """Increments training steps to run and blocks if training is too slow."""
```

Comment:
nit: I think this long comment works best when integrated into the method's docstring.

Reply:
done.